### PR TITLE
[DEM-850] Distance based classifier example

### PR DESCRIPTION
**Pull request by QFer**
* Branch with the files for the 'distance based classifier example' from pull request 39
* Changed the files to meet our code review standards and resolved Pep 8 violations.
* Renamed files
* installs the examples in the example directory venv/share/doc/quantuminspire/examples when you pip install.

### CODE REVIEW

File: docs/classifier_example/classification_v3_4_data_points.ipynb [hunk 1]
```diff
@@ -0,0 +1,793 @@
...
+   "outputs": [],
+   "source": [
+    "## Import external python file\n",
+    "import nbimporter\n",
+    "from data_preprocessing_class import DataPlotter  # for easier plotting \n",
+    "\n",
+    "# Import math functions \n",
+    "from math import acos, pi\n",
+    "from numpy import sign\n",
+    "get_bin = lambda x, n: format(int(x), 'b').zfill(n)\n",
```
Comment (qSaevar) [inline docs/classifier_example/classification_v3_4_data_points.ipynb]: Not very pep8 to assign a name to a lambda function.

File: docs/classifier_example/classification_v3_4_data_points.ipynb [hunk 2]
```diff
@@ -0,0 +1,793 @@
...
+    "    \"\"\" Gets the authentication for connecting to the Quantum Inspire API.\"\"\"\n",
+    "    if 'password' not in vars().keys():\n",
+    "        print('Enter email:')\n",
+    "        email = input()\n",
+    "        print('Enter password')\n",
+    "        password = getpass()\n",
+    "    return BasicAuthentication(email, password)\n",
+    "\n",
+    "# Remote Quantum-Inspire backend #\n",
+    "authentication = get_authentication()"
```
Comment (qSaevar) [inline docs/classifier_example/classification_v3_4_data_points.ipynb]: Should this not be more in line with the other notebooks:
```
token = load_account()
if token is not None:
    authentication = get_token_authentication(token)
else:
    if QI_EMAIL is None or QI_PASSWORD is None:
        ...
```
?

File: docs/classifier_example/classification_v3_4_data_points.ipynb [hunk 3]
```diff
@@ -0,0 +1,793 @@
...
+    "# Set-up a new connection with qi_backend:\n",
+    "def initialize_qi_backend(project_name = \"distance_based_classifier\"):\n",
+    "    compiler_engines = restrictedgateset.get_engine_list(one_qubit_gates=\"any\", two_qubit_gates=(CNOT, CZ, Toffoli))\n",
+    "    compiler_engines.extend([ResourceCounter(), ManualMapper(lambda x: x)])\n",
+    "    qi = QuantumInspireAPI(\n",
+    "        r'https://api.quantum-inspire.com/',\n",
+    "        authentication,\n",
+    "        project_name = project_name  # Set project name to save projects\n",
+    "    ) \n",
+    "    qi_backend = QIBackend(num_runs = 1, quantum_inspire_api=qi) # set num_runs = 1 for true probability distribution\n",
```
Comment (qSaevar) [inline docs/classifier_example/classification_v3_4_data_points.ipynb]: maybe it's just my machine, but this line does not fit in a cell

File: docs/classifier_example/classification_v3_4_data_points.ipynb [hunk 4]
```diff
@@ -0,0 +1,793 @@
...
...vd/frGsE2s0sB3P2qykYmIm3Z4ZHbzD4ILAH2M7MT0vJUYCfggorHJiJtGO20/P8AdwAXAn+Z1rYCv65yUCLSvtFOyw9w90cJN84s/doPOLricYlIm0Y7cr8V6APek6vXgdsrGZGIdMRo4f6qmU0HPjwegxGRzhkt3E44SmfV0trelYxIRDpih+F2973GayAi0lmjPRV2pbufbWaNdyrdxt2PqHRkItKW0U7LP5/+N39DTUQmuB0+Febuv0q/nQpcDtwC/BX6EAGRCa/s8tP/TVhfvgD4LvDNykYkIh1R9oUjWxof6gf8wMw+XtWARKQzRruhdlz67X+a2aeAlcChwK+a/ysRmQhGO3K/N/3vM4Rlp/uljzdVNiIR6YjRnudeXFQ3s1dXMxwR6ZSy78RyMXAWMB2YCTwC7F/huESkTWXvlh8P7AF8h3Bq/mRlIxKRjigb7qfdfROwi7sPEo7eEoGhoSGGhoZYv379tu8lDmXD/YSZnU64a34pMKvCMYlIB5R9nvvDhNPyGwnvgqrlqCITXNlw/x7wMSABHkLX3CITXtnT8uuBQeDThGBfV9mIRKQjyh65e9z9mvT7+83sXVUNSEQ6Y7Tlp0n67XozOwW4i7D89P9VPbDxUqtlH4UFePX8e8+ITEKjHbn/LvP9WcCZbH+bJRGZwEZbfnps43sz2w14PfBLd19f9cBEpD1lP5/7FGA14cMJ1pjZ+yodlYi0rezd8nOBXnc/GTiY8LSYiExgZcO91d1/C+DuLwAbqxuSiHRC2afCfmFmlxHerOEo4BfVDUlEOqHskXsJ8EtgYfrfD1U2IhHpiLJH7lvc/bjRm4nIRFE23M+a2UmEjxfaCuDuj1Q2KhFp26jhNrNZwF5A9h1P68BbqhqUiLRvtOWnZwOfALYAn3H3FeMyKhFp22g31E4DDJiPntsWmVRGC/dGd38pXW46fTwGJCKdUfapMAgvGBGRSWK0G2r7m9l3CcFufA+Au59W6chEpC2jhfvUzPdfr3IgItJZo73k88fjNRAR6axWrrlFZBJRuEUipXCLRKrs2vKWmNkU4GpgHuHjfpekH0PU2H4O2z/Y4DZ3v7iKcYh0s6qO3CcT3g75cOB84LLGBjPbG/jvwBHA4cBxZnZgReMQ6VpVhXsBsALA3dcAh2S2PQ78sbtvcfetwE7onV1EOq6S03LCBwU+l3m
```
Comment (qSaevar) [inline docs/classifier_example/classification_v3_4_data_points.ipynb]: Some blank lines missing, also below

File: docs/classifier_example/classification_v4_4_features.ipynb [hunk 5]
```diff
@@ -0,0 +1,884 @@
...
+    "def get_authentication():\n",
+    "    \"\"\" Gets the authentication for connecting to the Quantum Inspire API.\"\"\"\n",
+    "    if 'password' not in vars().keys():\n",
+    "        print('Enter email:')\n",
+    "        email = input()\n",
+    "        print('Enter password')\n",
+    "        password = getpass()\n",
+    "    return BasicAuthentication(email, password)\n",
+    "\n",
+    "authentication = get_authentication()\n",
```
Comment (qSaevar) [inline docs/classifier_example/classification_v4_4_features.ipynb]: Same comments as above, and also holds for the first notebook

File: docs/classifier_example/classification_v4_4_features.ipynb [hunk 6]
```diff
@@ -0,0 +1,884 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 5,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def cRy(q, angle, ctrlidx, idx):\n",
+    "    if len(q)<2:\n",
+    "        raise IndexError(\"Error, len quantum register must at least be 2.\")\n",
```
Comment (qSaevar) [inline docs/classifier_example/classification_v4_4_features.ipynb]: is this not a value error?

File: docs/classifier_example/classification_v4_4_features.ipynb [hunk 7]
```diff
@@ -0,0 +1,884 @@
...
...H3JoFdownHEjVVj8V4Mg+LQtCNmGCAIgYqlgMW5Si1SGyQH1JFqhahcICqVD2B1h9obRSBglalqRRMpCKKJkaapUIw4SFROfiDiAtqoGEXkkz/uHfI7OzMzszufDczz3xe5+zJzL3PfO+TTT733rnznZmeYrGIiMQzbW83ICJpKNwiQSncIkEp3CJBKdwiQSncIkHN2NsNSFpmVgTuA3YBRWAmsB042903JNre84CngVXu/rpWb0Mao3B3h6Pdfah0x8zeD3wKWJxwm88FDks4vtShcHcZM5sBvAR4rGzZRcDJZE/THgTOcfdfm9lfAH8P7CY78n/A3e80s+8Cn3b3r+ePH3U/dw3wLDP7MTAAXAK8CXgKeBQ43d0fTvhX7Xp6zt0d7jCze83s18AD+bJlAGZ2GnAwcJi7Hwp8E1iZ13yCLOgLgYuBP21im8uAJ/MxXwicB7w6H+s24DWT+ytJPTpyd4ej3X3IzF5FFt473P2RfN2fk50+bzAzgOlkz8sBvgKsMrNbgf8GPj7B7f8KuAe428xWA6vd/fYJjiUN0pG7i7j73cD7gC+Z2QH54unAx9z90PwouxB4bV5/EbAE2ACcDtyZP6YI9JQNvW+d7e4GjsrHeBS4yswmuqOQBincXcbdvwzcBVyVL/oWsNzMZuf3LweuN7MZZvYgMNPdPwecAxxiZr3Ab8l2ApjZQcAhVTa1E5huZj1mtoDsiv1md78i3/arU/z9ZA+dlnenc4F7zew4sufXLwLW5y9jPUR2sWunmZ0H3GhmT5NdVHunu4+Y2YeBa83sjcD97Dmil3uYbCfyE+AI4Ktkp/5PAE8C70n7V5QeveVTJCadlosEpXCLBJXkObeZTQOuBhYAI8Byd99atv6TZFdkH88Xnejuf0jRi0i3SnVB7SSgz90Xm9kiYAVwYtn6VwHHlU+JFJHWSnVavgRYA+Du68lfNoFnjurzgc+b2Voze2eiHkS6Wqoj92yg/DR7l5nNcPedwLPJ3rRwJdkEijvMbIO731s+wODgoC7jizRgYGCgp+qKYrHY8p9CoXBloVA4pez+L8tuTy8UCrPK7n+8UCj8VeUYGzZsKDZr06ZNbVPfTr00W69e2r+XkjwnVXOY6rR8LXA8QP6ce2PZugLwAzObbmb7kJ3C352oD5Guleq0fBWw1MzWkc1BXmZm5wNb3f1mM7sBWE
```
Comment (qSaevar) [inline docs/classifier_example/classification_v4_4_features.ipynb]: Here the error is just a string, above it is always IndexError

File: docs/classifier_example/data_preprocessing_class.py [hunk 8]
```diff
@@ -0,0 +1,551 @@
...
+    data.append("Sepal Length (cm)")
+    data.append("Sepal width (cm)")
+
+    print("Example for the iris test data based on first two features:")
+    plt, class1, class2 = pre_process_data(data, display_fig=True, axis_limits=[
+                                           [(-1, 8), (-1, 5)], [(-2.5, 2.5), (-2.5, 2.5)], [(-2.5, 2.5), (-2.5, 2.5)]])
+
+
+class DataPlotter:
+    def __init__(self):
```
Comment (qSaevar) [inline docs/classifier_example/data_preprocessing_class.py]: bit useless `__init__`

Comment (QFer) [reply docs/classifier_example/data_preprocessing_class.py]: removed it

**Comment by QFer** [comment]
* Renamed files

File: docs/classifier_example/classification_v3_4_data_points.ipynb [hunk 9]
```diff
@@ -0,0 +1,793 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "<a name=\"introduction\"></a>\n",
+    "# Introduction #\n",
+    "\n",
+    "\n",
+    "In the [previous part](link) of this notebook series on a distance-based classifier, we looked at how to implement a distance-based classifier on the quantum inspire using QASM code. We looked at the simplest possible case, that is: using two data points, each with two features, to assign a label (classify) to a random test point, see image. In this notebook we will extend the previous classifier, by increasing the number of data points to four. In this notebook we demonstrate how to use the projectQ framework to do this. \n",
```
[the remark this answers is not in the document]
Comment (QFer) [reply docs/classifier_example/classification_v3_4_data_points.ipynb]: In the first part of this...

File: docs/classifier_example/classification_v3_4_data_points.ipynb (hunk 1 above)
Comment (QFer) [reply docs/classifier_example/classification_v3_4_data_points.ipynb]: Removed it, not used in this note book

File: docs/classifier_example/classification_v3_4_data_points.ipynb (hunk 2 above)
Comment (QFer) [reply docs/classifier_example/classification_v3_4_data_points.ipynb]: Oke

File: docs/classifier_example/classification_v3_4_data_points.ipynb (hunk 3 above)
Comment (QFer) [reply docs/classifier_example/classification_v3_4_data_points.ipynb]: Depends on the width, I'll make some adjustments to keep the code the normal python length.

File: docs/classifier_example/classification_v4_4_features.ipynb (hunk 5 above)
Comment (QFer) [reply docs/classifier_example/classification_v4_4_features.ipynb]: Adjusted

File: docs/classifier_example/classification_v4_4_features.ipynb (hunk 6 above)
Comment (QFer) [reply docs/classifier_example/classification_v4_4_features.ipynb]: You're right, I'll change it to ValueError

File: docs/classifier_example/classification_v4_4_features.ipynb (hunk 7 above)
Comment (QFer) [reply docs/classifier_example/classification_v4_4_features.ipynb]: Also ValueError

File: docs/classifier_example/classification_example1_2_data_points.ipynb [hunk 10]
```diff
@@ -0,0 +1,1066 @@
...
+    "    \\begin{split}\n",
+    "        \\ket{\\psi_{\\bf\\tilde{x}}} & = 0.8670 \\ket{0} + 0.4984\\ket{1}, \\\\\n",
+    "        \\ket{\\psi_{\\bf x_1}} & = 0.9193 \\ket{0} + 0.3937\\ket{1},\\\\\n",
+    "        \\ket{\\psi_{\\bf x_2}} & = 0.1411 \\ket{0} + 0.9899\\ket{1}.\n",
+    "    \\end{split}\n",
+    "\\end{equation}\n",
+    "\n",
+    "Before we can run the actual algorithm we need to bring the system in the desired [initial state (equation 2)](#state) which can be obtain by applying the following combination of gates starting on $\\ket{0000}$. \n",
+    "\n",
+    "<font color=red>Note:</font> We originally interchanged qubit 1 and qubit 2 in the implementation on the quantum inspire. This means that in the images below, qubit $q[0]$ corresponds to the ancillary qubit and qubit $q[1]$ corresponds to the index register $\\ket{m}$.\n",
```
Comment (qSaevar) [inline docs/classifier_example/classification_example1_2_data_points.ipynb]: My machine hangs on this cell, does it work for you @QFer ?

File: docs/classifier_example/classification_example1_2_data_points.ipynb [hunk 11]
```diff
@@ -0,0 +1,1066 @@
...
+    "\n"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 1,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "## Import external python file\n",
```
Comment (qSaevar) [inline docs/classifier_example/classification_example1_2_data_points.ipynb]: I don't like this terminology, in this and the other notebooks.

File: docs/classifier_example/classification_example1_2_data_points.ipynb (hunk 10 above)
Comment (qSaevar) [reply docs/classifier_example/classification_example1_2_data_points.ipynb]: It worked in the end, after many many minutes

File: docs/classifier_example/classification_example3_4_features.ipynb [hunk 12]
```diff
@@ -0,0 +1,893 @@
...
+    "            print('Enter password')\n",
+    "            password = getpass()\n",
+    "        else:\n",
+    "            email, password = QI_EMAIL, QI_PASSWORD\n",
+    "        return get_basic_authentication(email, password)\n",
+    "\n",
+    "    \n",
+    "authentication = get_authentication()\n",
+    "# Temporairy alternative so that we can give our project a name:\n",
+    "QI._api = QuantumInspireAPI(QI_URL, authentication, project_name=\"Distance-based Classifier more features (NEW)\")\n",
```
Comment (qSaevar) [inline docs/classifier_example/classification_example3_4_features.ipynb]: access to private member, is there no other way to name the project?

Comment (QFer) [reply docs/classifier_example/classification_example3_4_features.ipynb]: We have [DEM-869] for this.

**Pull request merged**